### PR TITLE
Prover sessions based on free monads

### PR DESCRIPTION
**Pull request by loewenheim**
This branch changes the implementation of prover sessions to one based on free monads; see http://typelevel.org/cats/datatypes/freemonad.html for a pretty good explanation of what those are.

In practice, this means that you write a session as a mini-program inside gapt. For instance, you could have
```
val p = for {
  _ <- setLogic( "QF_UF" )
  _ <- declareSymbolsIn( s.elements )
  _ <- assert( s.map( identity, -_ ).elements.toList )
  _ <- checkSat
} yield ()
```
At this point, `p` has not been evaluated in any way. That only happens once you use a `Runner` to interpret it. For instance, `StackSessionRunner` interprets commands by pushing and popping formulas to and from a stack. On the other hand, `ExternalSMTLibSessionRunner` passes each command to an external process (`z3`, etc).

### CODE REVIEW

File: core/src/main/scala/at/logic/gapt/cutintro/improveSolutionLK.scala [hunk 1]
```diff
@@ -50,39 +51,53 @@ object improveSolutionLK {
...
+        val clauses = for ( inst <- groundInstances; clause <- cnf ) yield inst( clause.toDisjunction )
+        for {
+          _ <- init
+          _ <- assert( clauses.toList )
+          sat <- checkSat
+        } yield sat
+      }
 
-      checkSolution( CNFp( start ).map { _.distinct.sortBy { _.hashCode } } )
+      if ( !prover.runSession( checkSolutionProgram ) ) {
```
Comment (gebner) [inline core/src/main/scala/at/logic/gapt/cutintro/improveSolutionLK.scala]: This line starts a new Z3 (etc.) instance every time it is executed.  Kind of the whole point behind the current interactive session is that we only need to start the solver once--and not only that, if we want to do many queries of the form `Γ ∪ Δᵢ` with a fixed `Γ`, then we only need to send the `Γ` once.  (We do not know the full list of the `Δᵢ` in advance.)  This is a very nice performance boost, at least for Z3 and CVC4, which support this kind of interactive operation.

Comment (loewenheim) [inline core/src/main/scala/at/logic/gapt/cutintro/improveSolutionLK.scala]: Fair point, I'll see about refactoring this.

**Comment by gebner** [comment]
:tada: This is really nice.  I think we should merge it.

As I already told you over lunch, I think the Session monad might be the right interface to use for all the external provers.

**Comment by loewenheim** [comment]
Thanks for the compliment. Your reformatting is also much more readable than my version. If you say we merge it, I have no objections.

**Pull request merged**